File: 240922/고대 문명 유적 탐사/ancient-ruin-exploration.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader br;
	static StringBuilder sb;
	static StringTokenizer st;

	static int[] dirX = { -1, 0, 0, 1 };
	static int[] dirY = { 0, 1, -1, 0 };
	static int K, M;
	static int[][] grid;
	static Queue<Integer> nextItems;

	public static void main(String[] args) throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));
		sb = new StringBuilder();
		st = new StringTokenizer(br.readLine());
		K = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		grid = new int[5][5];
		for (int i = 0; i < 5; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < 5; j++) {
				grid[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		nextItems = new ArrayDeque<>();
		st = new StringTokenizer(br.readLine());
		while (st.hasMoreTokens()) {
			nextItems.add(Integer.parseInt(st.nextToken()));
		}
		for (int k = 0; k < K; k++) {
			final Coordinate coordinate = findMaxValueCoordinate();
			if (coordinate.value == 0) {
				break;
			}
			updateGrid(coordinate.d, coordinate.sx, coordinate.sy);
			int removedItemCount = removeUsedItems();
			int itemValueSum = 0;
			while (removedItemCount > 0) {
				itemValueSum += removedItemCount;
				fillNextItemsToEmptySpace();
				removedItemCount = removeUsedItems();
			}
			sb.append(itemValueSum)
					.append(" ");
		}
		System.out.println(sb);
	}

	private static void rotate90(final int sx, final int sy) {
		final int[][] temp = new int[5][5];
		for (int x = sx; x < sx + 3; x++) {
			for (int y = sy; y < sy + 3; y++) {
				final int ox = x - sx;
				final int oy = y - sy;
				final int rx = oy;
				final int ry = 3 - ox - 1;
				temp[sx + rx][sy + ry] = grid[x][y];
			}
		}
		for (int x = sx; x < sx + 3; x++) {
			for (int y = sy; y < sy + 3; y++) {
				grid[x][y] = temp[x][y];
			}
		}
	}

	private static Coordinate findMaxValueCoordinate() {
		final List<Coordinate> coordinates = new ArrayList<>();
		for (int sx = 0; sx < 3; sx++) {
			for (int sy = 0; sy < 3; sy++) {
				for (int d = 0; d < 4; d++) {
					rotate90(sx, sy);
					if (d == 3) {
						break;
					}
					final boolean[][] visited = new boolean[5][5];
					int value = 0;
					for (int x = 0; x < 5; x++) {
						for (int y = 0; y < 5; y++) {
							if (!visited[x][y]) {
								value += bfs1(x, y, visited);
							}
						}
					}
					coordinates.add(new Coordinate(value, d, sx, sy));
				}
			}
		}
		Collections.sort(coordinates);
		return coordinates.get(0);
	}

	private static int bfs1(final int sx, final int sy, final boolean[][] visited) {
		final Queue<Pos> queue = new ArrayDeque<>();
		queue.add(new Pos(sx, sy));
		visited[sx][sy] = true;
		final int number = grid[sx][sy];
		int count = 1;
		while (!queue.isEmpty()) {
			final Pos pos = queue.poll();
			for (int d = 0; d < 4; d++) {
				final int nx = pos.x + dirX[d];
				final int ny = pos.y + dirY[d];
				if (0 <= nx && nx < 5 && 0 <= ny && ny < 5 && !visited[nx][ny] && grid[nx][ny] == number) {
					queue.add(new Pos(nx, ny));
					visited[nx][ny] = true;
					count++;
				}
			}
		}
		if (count < 3) {
			return 0;
		}
		return count;
	}

	private static void updateGrid(final int dir, final int sx, final int sy) {
		for (int d = 0; d < dir + 1; d++) {
			rotate90(sx, sy);
		}
	}

	private static int removeUsedItems() {
		int count = 0;
		final boolean[][] visited = new boolean[5][5];
		for (int x = 0; x < 5; x++) {
			for (int y = 0; y < 5; y++) {
				count += bfs2(x, y, visited);
			}
		}
		return count;
	}

	private static int bfs2(final int sx, final int sy, final boolean[][] visited) {
		final Queue<Pos> sameNumberPos = new ArrayDeque<>();
		sameNumberPos.add(new Pos(sx, sy));
		final Queue<Pos> queue = new ArrayDeque<>();
		queue.add(new Pos(sx, sy));
		visited[sx][sy] = true;
		final int number = grid[sx][sy];
		while (!queue.isEmpty()) {
			final Pos pos = queue.poll();
			for (int d = 0; d < 4; d++) {
				final int nx = pos.x + dirX[d];
				final int ny = pos.y + dirY[d];
				if (0 <= nx && nx < 5 && 0 <= ny && ny < 5 && !visited[nx][ny] && grid[nx][ny] == number) {
					final Pos nextPos = new Pos(nx, ny);
					queue.add(nextPos);
					sameNumberPos.add(nextPos);
					visited[nx][ny] = true;
				}
			}
		}
		if (sameNumberPos.size() < 3) {
			return 0;
		}
		int count = 0;
		while (!sameNumberPos.isEmpty()) {
			final Pos pos = sameNumberPos.poll();
			grid[pos.x][pos.y] = 0;
			count++;
		}
		return count;
	}

	private static void fillNextItemsToEmptySpace() {
		for (int y = 0; y < 5; y++) {
			for (int x = 4; x >= 0; x--) {
				if (grid[x][y] == 0) {
					grid[x][y] = nextItems.poll();
				}
			}
		}
	}

	private static class Pos {

		final int x;
		final int y;

		public Pos(final int x, final int y) {
			this.x = x;
			this.y = y;
		}
	}

	private static class Coordinate implements Comparable<Coordinate> {

		final int value;
		final int d;
		final int sx;
		final int sy;

		public Coordinate(final int value, final int d, final int sx, final int sy) {
			super();
			this.value = value;
			this.d = d;
			this.sx = sx;
			this.sy = sy;
		}

		@Override
		public int compareTo(final Coordinate other) {
			// TODO Auto-generated method stub
			if (this.value != other.value) {
				return other.value - this.value;
			}
			if (this.d != other.d) {
				return this.d - other.d;
			}
			if (this.sy != other.sy) {
				return this.sy - other.sy;
			}
			return this.sx - other.sx;
		}
	}
}

File: 250306/고대 문명 유적 탐사/ancient-ruin-exploration.py
import heapq
from collections import deque


def rotate_90(sx, sy):
    global grid
    temp = [[0] * 5 for _ in range(5)]
    for x in range(sx, sx + 3):
        for y in range(sy, sy + 3):
            ox = x - sx
            oy = y - sy
            rx = oy
            ry = 3 - ox - 1
            temp[sx + rx][sy + ry] = grid[x][y]
    for x in range(sx, sx + 3):
        for y in range(sy, sy + 3):
            grid[x][y] = temp[x][y]


def find_max_value_coordinate():
    candidates = []
    for sx in range(3):
        for sy in range(3):
            for d in range(4):
                rotate_90(sx, sy)
                if d == 3:
                    break
                visited = [[False] * 5 for _ in range(5)]
                cur_value = 0
                for x in range(5):
                    for y in range(5):
                        if not visited[x][y]:
                            cur_value += bfs1(x, y, visited)
                heapq.heappush(candidates, (-cur_value, d, sy, sx))
    return heapq.heappop(candidates)


def bfs1(x, y, visited):
    queue = deque([(x, y)])
    visited[x][y] = True
    number = grid[x][y]
    cnt = 1
    while queue:
        x, y = queue.popleft()
        for dx, dy in dirs:
            nx = x + dx
            ny = y + dy
            if 0 <= nx < 5 and 0 <= ny < 5 and not visited[nx][ny] and grid[nx][ny] == number:
                queue.append((nx, ny))
                visited[nx][ny] = True
                cnt += 1
    if cnt < 3:
        return 0
    return cnt


def update_grid(d, sx, sy):
    global grid
    for _ in range(d + 1):
        rotate_90(sx, sy)


def remove_used_items():
    cnt = 0
    visited = [[False] * 5 for _ in range(5)]
    for x in range(5):
        for y in range(5):
            if not visited[x][y]:
                cnt += bfs2(x, y, visited)
    return cnt


def bfs2(x, y, visited):
    global grid
    queue = deque([(x, y)])
    visited[x][y] = True
    number = grid[x][y]
    same_numbers = deque([(x, y)])
    while queue:
        x, y = queue.popleft()
        for dx, dy in dirs:
            nx = x + dx
            ny = y + dy
            if 0 <= nx < 5 and 0 <= ny < 5 and not visited[nx][ny] and grid[nx][ny] == number:
                queue.append((nx, ny))
                same_numbers.append((nx, ny))
                visited[nx][ny] = True
    if len(same_numbers) < 3:
        return 0
    cnt = 0
    while same_numbers:
        x, y = same_numbers.popleft()
        grid[x][y] = 0
        cnt += 1
    return cnt


def fill_next_items_to_empty_space():
    global grid
    for y in range(5):
        for x in range(4, -1, -1):
            if grid[x][y] == 0:
                grid[x][y] = next_items.popleft()


K, M = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(5)]
next_items = deque(list(map(int, input().split())))
dirs = ((-1, 0), (0, 1), (0, -1), (1, 0))
for _ in range(K):
    value, d, sy, sx = find_max_value_coordinate()
    if value == 0:
        break
    update_grid(d, sx, sy)
    cur_value = remove_used_items()
    item_values_sum = 0
    while cur_value > 0:
        item_values_sum += cur_value
        fill_next_items_to_empty_space()
        cur_value = remove_used_items()
    print(item_values_sum, end=' ')
